uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/absence-tab/components/annual-entitlements.js

```diff
@@ -0,0 +1,177 @@
+
+      var vm = Object.create(this);
+      var contacts = [];
+      var allEntitlements = [];
+
+      vm.absencePeriods = [];
+      vm.absenceTypes = [];
+      vm.loading = { absencePeriods: true };
+      vm.editEntitlementsPageUrl = getEditEntitlementsPageURL(vm.contactId);
+
+      (function init () {
+        return $q.all([
+          loadAbsenceTypes(),
+          loadEntitlements()
+        ])
+        .then(function () {
+          return loadCommentsAuthors();
+        })
+        .then(function () {
+          return loadAbsencePeriods();
+        })
+        .finally(function () {
+          vm.loading.absencePeriods = false;
+        });
+      })();
+
+      /**
+       * Shows a comment to the entitlement
+       *
+       * @param {object} comment
+       */
+      vm.showComment = function (comment) {
+        /*
+         * @TODO There is no support for footer in CRM.alert at the moment.
+         * This code should be refactored as soon as CRM.alert supports footer.
+         * At the moment the footer is constructed via rich HTML directly via body text
+         */
+        var text = comment.message +
+          '<br/><br/><strong>Last updated:' +
+          '<br/>By: ' + comment.author_name +
+          '<br/>Date: ' + moment(comment.date).format('DD/M/YYYY HH.mm') +
+          '</strong>';
+
+        CRM.alert(text, 'Calculation comment:', 'error');
+      };
+
+      /**
+       * Loads absence periods
+       *
+       * @return {Promise}
+       */
+      function loadAbsencePeriods () {
+        return AbsencePeriod.all().then(setAbsencePeriodsProps);
+      }
+
+      /**
+       * Loads absence types
+       *
+       * @return {Promise}
+       */
+      function loadAbsenceTypes () {
+        return AbsenceType.all()
+          .then(function (data) {
+            vm.absenceTypes = data;
+          });
+      }
+
+      /**
+       * Loads entitlements comments authors
+       *
+       * @return {Promise}
+       */
+      function loadCommentsAuthors () {
+        var authorsIDs = _.uniq(_.map(allEntitlements, function (entitlement) {
+          return entitlement.comment_author_id;
+        }));
+
+        return Contact.all({ id: { 'IN': authorsIDs } })
+          .then(function (data) {
+            contacts = _.indexBy(data.list, 'contact_id');
+          });
+      }
+
+      /**
+       * Loads entitlements
+       *
+       * @return {Promise}
+       */
+      function loadEntitlements () {
+        return Entitlement.all({ contact_id: vm.contactId })
+          .then(function (data) {
+            allEntitlements = data;
+          });
+      }
+
+      /**
+       * Processes entitlements from data and sets them to the controller
+       *
+       * @param {object} absencePeriods
+       */
+      function setAbsencePeriodsProps (absencePeriods) {
+        // Get all periods as per entitlements
+        var periods = _.uniq(_.map(allEntitlements, function (entitlement) {
```

Comment:
L125 - get IDs of only periods the user has entitlements for
L130 - filter only those periods that user has entitlements for
L133 - sort periods for better UX